blackjax/adaptation/chees_adaptation.py

```diff
@@ -0,0 +1,426 @@
+    mean of the chain's acceptance probabilities to tune the step size.
+
+    Parameters
+    ----------
+    jitter_generator
+        Optional function that generates a value in [0, 1] used to jitter the trajectory
+        lengths given a PRNGKey, used to propose the number of integration steps. If None,
+        then a quasi-random Halton is used to jitter the trajectory length.
+    next_random_arg_fn
+        Function that generates the next `random_generator_arg` from its previous value.
+    optim
+        Optax compatible optimizer, which conforms to the `optax.GradientTransformation` protocol.
+
+    Returns
+    -------
+    init
+        Function that initializes the warmup.
+    update
+        Function that moves the warmup one step.
+
+    """
+
+    da_init, da_update, _ = dual_averaging.dual_averaging()
+
+    def compute_parameters(
+        proposed_positions: ArrayLikeTree,
+        proposed_momentums: ArrayLikeTree,
+        initial_positions: ArrayLikeTree,
+        acceptance_probabilities: Array,
+        is_divergent: Array,
+        initial_adaptation_state: ChEESAdaptationState,
+    ):
+        """Compute values for the parameters based on statistics collected from
+        multiple chains.
+
+        Parameters
+        ----------
+        proposed_positions:
+            A PyTree that contains the position proposed by the HMC algorithm of
+            every chain (proposal that is accepted or rejected using MH).
+        proposed_momentums:
+            A PyTree that contains the momentum variable proposed by the HMC algorithm
+            of every chain (proposal that is accepted or rejected using MH).
+        initial_positions:
+            A PyTree that contains the initial position at the start of the HMC
+            algorithm of every chain.
+        acceptance_probabilities:
+            Metropolis-Hastings acceptance probabilty of proposals of every chain.
+        initial_adaptation_state:
+            ChEES adaptation step used to generate proposals and acceptance probabilities.
+
+        Returns
+        -------
+        New values of the step size and trajectory length of the jittered HMC algorithm.
+
+        """
+        (
+            step_size,
+            step_size_ma,
+            trajectory_length,
+            trajectory_length_ma,
+            da_state,
+            optim_state,
+            random_generator_arg,
+        ) = initial_adaptation_state
+
+        harmonic_mean = 1.0 / jnp.mean(
+            1.0 / acceptance_probabilities, where=~is_divergent
+        )
+        da_state_ = da_update(da_state, 0.651 - harmonic_mean)
+        step_size_ = jnp.exp(da_state_.log_x)
+        new_step_size, new_da_state = jax.lax.cond(
+            jnp.isfinite(step_size_),
+            lambda _: (step_size_, da_state_),
+            lambda _: (step_size, da_state),
+            None,
+        )
+        new_step_size_ma = 0.9 * step_size_ma + 0.1 * new_step_size
```

Comment:
The moving average in tfp looks different (same comment applies below): https://github.com/tensorflow/probability/blob/bc2e0e0d0969d2b5e7cb0a8fbaaf93ba1e0adfb2/tensorflow_probability/python/experimental/mcmc/gradient_based_trajectory_length_adaptation.py#L1004-L1010

Reply:
They do a couple of things differently, like having an [extra fixed jitter value](https://github.com/tensorflow/probability/blob/bc2e0e0d0969d2b5e7cb0a8fbaaf93ba1e0adfb2/tensorflow_probability/python/experimental/mcmc/gradient_based_trajectory_length_adaptation.py#L786C28-L787C68). 

They update the moving average for the trajectory length using a weighted geometric mean with the weight depending on the iteration number instead of the weighted mean with fixed weights used in [Algorithm 1](http://proceedings.mlr.press/v130/hoffman21a/hoffman21a.pdf).

Also, this script only seems to update the trajectory length; I can't find updates for the step size. In Algorithm 1, authors use the same weighted mean with fixed weights used for the trajectory length. 

The question is if we want to stick to Algo 1 or use more of the methods in tfp. We are already using their unshuffled generation of a [halton sequence](https://github.com/tensorflow/probability/blob/bc2e0e0d0969d2b5e7cb0a8fbaaf93ba1e0adfb2/tensorflow_probability/python/experimental/mcmc/gradient_based_trajectory_length_adaptation.py#L924).

Reply:
similar weighted geometric mean for step size: https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/mcmc/dual_averaging_step_size_adaptation.py#L460C5-L460C27

Reply:
> They update the moving average for the trajectory length using a weighted geometric mean with the weight depending on the iteration number instead of the weighted mean with fixed weights used in [Algorithm 1](http://proceedings.mlr.press/v130/hoffman21a/hoffman21a.pdf).

I think it make sense to use the weighted geometric mean with the weight depending on the iteration number (TFP current), as it should be better than a fixed weight

The step size adaptation in TFP is also dual averaging, the implementation here using blackjax dual_averaging is equivalent.

Reply:
We should also add the [extra fixed jitter value](https://github.com/tensorflow/probability/blob/bc2e0e0d0969d2b5e7cb0a8fbaaf93ba1e0adfb2/tensorflow_probability/python/experimental/mcmc/gradient_based_trajectory_length_adaptation.py#L786C28-L787C68). It's pretty straightfoward to do, and the default value is 1 (no jitter)